src/main/kotlin/uk/gov/justice/digital/hmpps/prisonertonomisupdate/csip/CSIPReconciliationService.kt

```diff
@@ -51,20 +55,33 @@ class CSIPReconciliationService(
       .also { log.info("Page requested: $page, with ${it.size} active prisoners") }
 
   suspend fun checkOpenCSIPsMatch(prisonerId: PrisonerIds): MismatchCSIPs? = runCatching {
-    val nomisCSIPs = doApiCallWithRetries { nomisCSIPApiService.getCSIPsForReconciliation(prisonerId.offenderNo) }.offenderCSIPs.size
-    val dpsCSIPs = doApiCallWithRetries { dpsCSIPApiService.getCSIPsForPrisoner(prisonerId.offenderNo) }.size
+    val nomisCSIPList = doApiCallWithRetries { nomisCSIPApiService.getCSIPsForReconciliation(prisonerId.offenderNo) }.offenderCSIPs
+    val dpsCSIPList = doApiCallWithRetries { dpsCSIPApiService.getCSIPsForPrisoner(prisonerId.offenderNo) }
 
-    val mismatchCSIP = dpsCSIPs - nomisCSIPs
-    return if (mismatchCSIP != 0) {
-      MismatchCSIPs(offenderNo = prisonerId.offenderNo, nomisCSIPCount = nomisCSIPs, dpsCSIPCount = dpsCSIPs).also { mismatch ->
+    val dpsCSIPs = dpsCSIPList.map { it.toCSIPSummary() }.toSet()
+    val nomisCSIPs = nomisCSIPList.map { it.toCSIPSummary() }.toSet()
+
+    val mismatchCount = nomisCSIPList.size - dpsCSIPList.size
+    val missingFromNomis = dpsCSIPs - nomisCSIPs
+    val missingFromDps = nomisCSIPs - dpsCSIPs
+    return if (mismatchCount != 0 || missingFromNomis.isNotEmpty() || missingFromDps.isNotEmpty()) {
+      MismatchCSIPs(
+        offenderNo = prisonerId.offenderNo,
+        dpsCSIPCount = dpsCSIPList.size,
+        nomisCSIPCount = nomisCSIPList.size,
+        missingFromDps = missingFromDps,
+        missingFromNomis = missingFromNomis,
+      ).also { mismatch ->
         log.info("CSIP Mismatch found  $mismatch")
```

Comment:
Do you need to `log` and track the event?  Why do both - we're looking to keep our Log Analytics costs down so don't know if we need both?

Reply:
Yeah totally get that - in theory we should only get a few mismatches - if any, but I can totally see your point.